### ISSUE TITLE
Possible new auth flow?

### ISSUE THREAD

**Issue by IdealText** [eidorb/ubank#6]
Python version: 3.11
OS: MacOS 15.2
fetched from master, v1.1.0

I have recently signed up to ubank and looking for a way to get my transaction history and was trying out this project. I hit a few issues with the sign up process.

In the response to `https://api.ubank.com.au/app/v1/challenge/otp` I was getting the error:
```
{'message': 'Invalid request: flowId not valid uuid', 'stackError': {'code': 'FEA-400'}}
``` 

I looked at the response payload from `https://api.ubank.com.au/app/v1/challenge/password` and saw it included the `flowId`. I included this in the json of the request to `otp` and got back a payload, but it was different to what was expected. It was the following

```
{
  "accessToken": "removed",
  "sessionToken": "removed",
  "expiresIn": 600,
  "userId": "removed",
  "username": "removed",
  "mobileNumber": "removed",
  "email": "removed",
  "refreshToken": "removed",
  "refreshExpiresIn": 600,
  "type": "bearer"
}
```

Missing `authKey`. 

I played around further to see what may be going on, I noticed that when I login to the app on my phone I am unable to have a persistent login. If I use username + password + SMS, then each time I access the app I have to reauthenticate. This is then pushing me to use a passkey in the UI.

Is it possible that ubank have moved to remove long lived logins, at least for new users?

**Comment by jakepronger** [comment]
I've noticed this myself and don't think it evens works now even with a device already registered. Will look into this later when I get more time, but it does appear that they've changed some internal auth things to make it more complicated. Maybe there is a way to automate or generate a passkey instead and replicate the same behavior with this new auth method.

**Comment by eidorb** [comment]
Yes enrolling fails for me after the OTP step too. But using my existing credentials still works. 

I will look into the passkey route as that should be more future proof. A fun challenge too I'm sure.

**Comment by jakepronger** [comment]
> Yes enrolling fails for me after the OTP step too. But using my existing credentials still works.
> 
> I will look into the passkey route as that should be more future proof. A fun challenge too I'm sure.

Yes. Been looking at this now and notice the request doesn't return the authKey needed to access the account. Been debugging with this but I wanted to know whether you used any tools to intercept any https traffic with the app itself when you started out this project?

**Comment by jakepronger** [comment]
I need to understand the structure of the request referencing this section of code

`            # Authenticate with long life token.
            response = client.post(
                "https://api.ubank.com.au/app/v1/long-life-token/login",
                json={
                    #"authKey": device.auth_key,
                    "deviceUuid": device.device_id,
                    "secret": device.secret,
                    "token": device.token,
                    "username": device.username,
                },
            )`

**Comment by jakepronger** [comment]
I can use inspect element and connect my mobile to track all of the requests and see it's contents. I've done some things but the authKey is missing from one of the request's responses. Does anyone know the format of the authKey or does it look like the normal token.

**Comment by jakepronger** [comment]
Is it the same value as 'token' or should be different?

**Comment by eidorb** [comment]
They are different. You get them at various stages during initial device enrolment:

https://github.com/eidorb/ubank/blob/eced2e794f140cf1af0fb6b862b5e62055ee4425/ubank.py#L274
https://github.com/eidorb/ubank/blob/eced2e794f140cf1af0fb6b862b5e62055ee4425/ubank.py#L279

And when you login, I think they update with new values.

**Comment by jakepronger** [comment]
They don't send an authKey in the otp response anymore. Nothing new in the headers either.

This is all we are given in the otp response:

{
    "accessToken": "",
    "sessionToken": "",
    "expiresIn": 600,
    "userId": "",
    "username": "",
    "mobileNumber": "",
    "email": "",
    "refreshToken": "",
    "refreshExpiresIn": 600,
    "type": "bearer"
}

**Comment by jakepronger** [comment]
Not sure what can be done about this but still looking into it.

**Comment by eidorb** [comment]
Yeah I think this method is busted for enrolling new devices. They are forcing SMS OTP for every login with password auth.

Interestingly my old auth key and token still works ok.

**Comment by jakepronger** [comment]
Very interesting. If only there was a way to find that auth key that they are not providing us. My old auth key doesn't work but I'm pretty sure that's because of something I may have done, I don't remember.

`error response:  {'message': '', 'stackError': {'detailMessage': {}, 'message': 'Invalid device binding.  Rebind device via device activation', 'code': 'IDENTITY-LOGIN_FAILED_INVALID_BINDING'}}`

**Comment by jakepronger** [comment]
The flowId is an easy fix just by providing it in the otp_response post.

```py
        # Authenticate with second factor: a security code sent to mobile.
        otp_response = client.post(
            url="https://api.ubank.com.au/app/v1/challenge/otp",
            # Set parameters returned in previous response.
            params={
                "nonce": response.json()["nonce"],
                "state": response.json()["state"],
                "session": response.json()["session"],
            },
            json={
                "deviceId": device_id,
                "hashedPin": hashed_pin,
                "flowId": flowId,
                # Prompt interactively for security code.
                "otpValue": input(
                    f'Enter security code sent to {response.json()["maskedMobileNumber"]}: '
                ),
            },
            # Set access and auth token headers from previous response.
            headers={
                "x-access-token": response.json()["accessToken"],
                "x-auth-token": response.json()["accessToken"],
            },
        )
```

**Comment by jakepronger** [comment]
It does look like we will need to change this up and use some kind of passkey authentication. They use FIDO2 password less passkeys which I'm just looking into now and seeing what I can try and do with this.

**Comment by eidorb** [comment]
Cool! I've started to look into it too. I'm able to create virtual passkeys without a browser or hardware device using this https://github.com/bodik/soft-webauthn

**Comment by jakepronger** [comment]
Very interesting. I will look at that too.

**Comment by eidorb** [comment]
I've managed to add a new passkey with Python. There's a bit more to do, but a promising step.

![Image](https://github.com/user-attachments/assets/ef719f5d-3b88-458a-8e58-38dbf8f99d95)

**Comment by jakepronger** [comment]
Very cool. I got called into work so I guess i will have to look into this tonight.

**Comment by eidorb** [comment]
I've [released](https://pypi.org/project/ubank/2.0.0rc0/) a package with passkey support. The README has been updated with new instructions.

It's pre-release so you'll have to specify the version when installing: `pip install ubank==2.0.0rc0`.

**Comment by eidorb** [comment]
New pre-release changes the passkey serialization format: `pip install ubank==2.0.0rc1`

**Comment by jakepronger** [comment]
Woah. You did good stuff. I'm going to look into doing CommBank as I want to automate transactions between my two accounts but that passkey thing is so good.

**Comment by eidorb** [comment]
![Image](https://github.com/user-attachments/assets/6c3d4cc3-d7d0-4715-a771-cf0cb59e6c8a)

😆

**Comment by jakepronger** [comment]
Nice. I need to look into commbank but can't see the requests easily.
